db/migrate/20230719100256_migrate_recurring_count_and_unique_count_aggregation.rb

```diff
@@ -0,0 +1,151 @@
+  class Charge < ApplicationRecord; end
+  class Customer < ApplicationRecord; end
+  class Event < ApplicationRecord; end
+  class Plan < ApplicationRecord; end
+  class Subscription < ApplicationRecord; end
+
+  def change
+    reversible do |dir|
+      dir.up do
+        # Create quantified event object for all unique count events
+        sql = <<-SQL
+          SELECT
+            events.id AS event_id,
+            events.code AS code,
+            events.organization_id AS organization_id,
+            events.customer_id AS customer_id,
+            events.properties AS properties,
+            events.timestamp AS event_timestamp,
+            subscriptions.external_id AS subscription_external_id
+          FROM events
+          INNER JOIN subscriptions ON subscriptions.id = events.subscription_id
+          INNER JOIN billable_metrics ON billable_metrics.code = events.code
+          WHERE events.deleted_at IS NULL AND billable_metrics.aggregation_type = 3
+          ORDER BY event_timestamp ASC;
+        SQL
+
+        ApplicationRecord.connection.select_all(sql).each_with_object({}) do |row, _result|
+          matching_billable_metric = BillableMetric.find_by(
+            code: row['code'],
+            organization_id: row['organization_id'],
+          )
+
+          existing_quantified_event =
+            QuantifiedEvent.where(
+              customer_id: row['customer_id'],
+              billable_metric_id: matching_billable_metric.id,
+              external_subscription_id: row['subscription_external_id'],
+              external_id: JSON.parse(row['properties'])[matching_billable_metric.field_name.to_s],
+            ).where(removed_at: nil).any?
+
+          # There can only be one quantified event for certain external_id which guarantees uniqueness
+          next if existing_quantified_event
+
+          quantified_event = QuantifiedEvent.create!(
+            customer_id: row['customer_id'],
+            billable_metric_id: matching_billable_metric.id,
+            external_subscription_id: row['subscription_external_id'],
+            external_id: JSON.parse(row['properties'])[matching_billable_metric.field_name.to_s],
+            properties: JSON.parse(row['properties']),
+            added_at: row['event_timestamp'],
+          )
+
+          event = Event.find_by(id: row['event_id'])
+          event.quantified_event_id = quantified_event.id
+          event.save!
+        end
+
+        # All charges that are related to recurring_count_agg should be prorated = true
+        sql = <<-SQL
+          SELECT
+            charges.id AS charge_id
+          FROM charges
+          INNER JOIN billable_metrics ON billable_metrics.id = charges.billable_metric_id
+          WHERE billable_metrics.aggregation_type = 4;
+        SQL
+        ApplicationRecord.connection.select_all(sql).each_with_object({}) do |row, _result|
+          charge = Charge.find_by(id: row['charge_id'])
```

Comment:
We should avoid as much as possible using rails models inside migration since we add a lot of issues with this in the past.

For example this one could be written in pure SQL:

```ruby
execute <<-SQL
  WITH recurring_count_charges AS (
    SELECT charges.id AS charge_id
    FROM charges
      INNER JOIN billable_metrics ON billable_metrics.id = charges.billable_metric_id
    WHERE billable_metrics.aggregation_type = 4 -- recurring_count_agg
  )

  UPDATE charges
  SET prorated = true
  FROM recurring_count_charges
  WHERE recurring_count_charges.charge_id = charges.id
SQL
```